WorldProcessor prototype - version 0.0.1.0

N0rator opened the pull request.

N0rator committed:
world-iterator: update gitignore files

N0rator committed:
world-iterator: create initial project

N0rator committed:
world-iterator: updated iterateworld function contract

N0rator committed:
world-iterator: implemented world iteration webapi service with clean architecture

N0rator committed:
world-iterator: renamed project to WorldProcessor

Max-Dov commented:
closing due to no progress

N0rator committed:
world-processor: add world generation functionality, fix bugs

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+# 🏭 World Iterator
```
```suggestion
# 🏭 World Processor
```
Везде называется как процессор, кроме названия

N0rator replied:
Принято, добавлено.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+# 🏭 World Iterator
+WebApi service that provide functionality to generate and iterate world.
```
```suggestion
Web API service that provides functionality to generate and iterate world.
```

N0rator replied:
Принято, добавлено.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+# 🏭 World Iterator
+WebApi service that provide functionality to generate and iterate world.
+
+## Usage
```
If it's Web API, better just call usage an API
```suggestion
## API
```

N0rator replied:
Принято, добавлено.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+WebApi service that provide functionality to generate and iterate world.
+
+## Usage
+WorldProcessor have 2 endpoints:
```
```suggestion
World Processor has 2 endpoints:
```

N0rator replied:
Принято, добавлено.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Usage
+WorldProcessor have 2 endpoints:
+1) **url:port/WorldProcessing/Generate** - action for generation initial world.
```
```suggestion
### Generate World

`POST https://localhost:3000/WorldProcessing/Generate`

Request body:
```
* It is not numeric list (order is not important) and list heading should be headers, so it can be seeing in "heading" if readme had one and could be refered via link (e.g. "/readme.md#generate")

N0rator replied:
Принято, добавлено.

Max-Dov commented on WorldProcessor/README.md:
````diff
@@ -0,0 +1,38 @@
+1) **url:port/WorldProcessing/Generate** - action for generation initial world.
+	It accepts only POST request.
+	request description:
+	```
````
```suggestion
	```language
```
If possible, make it language for nice synthax highlight, otherwise I recommend making model (aka "request description") TypeScript interface or JSON-line object as it's self explanatory and can be referenced via notes/contracts.md.

N0rator replied:
Added typescript as language

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+
+##  Compiled application
+Latest compiled application is located in `Release build` directory.
```
Make it link to github exact file on master branch, otherwise it doesn't mean latest, it means "this branch executable."

N0rator replied:
Added link to application directory, becuase it is not just single file

Max-Dov commented on WorldProcessor/README.md:
````diff
@@ -0,0 +1,38 @@
+	```
+
+
+##  Compiled application
````
```suggestion
##  Executable file
```
Optional task, but I think "compiled" and "application" is too broad description for simple ".exe" file.

N0rator replied:
I prefer more "Latest application build", because it is not a single file

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+##  Compiled application
+Latest compiled application is located in `Release build` directory.
+To use it, just run `WorldProcessor.WebApi.exe` executable file and find application address and port in console.
+![CLI Preview](./readme-preview-editor.png "CLI Preview")
```
lmao, did you even preview your readme?

```suggestion
![CLI Preview](./readme-preview-cli.png "CLI Preview")
```

Also logs make no sense to me as screenshot has both :5000 and :5001 as ports, so what is actual port??

Can you like.. clean console when app starts and print simple message "World Processor started! Address: https://localhost:1234/"?

N0rator replied:
Applicaiton listens both ports. 5000 for http and 5001 for https requests. This information will be added to readme file.
I don't know how to change default cli output) If it is important i can investigate this issue.
However image will be removed.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Development
+For development install .NET SDK - https://dotnet.microsoft.com/en-us/download.
+Use IDE that support development on .NET like [Visual Studio](https://visualstudio.microsoft.com/ru/) or [Intellij Rider](https://www.jetbrains.com/ru-ru/rider/).
```
Make it english links if readme is in english.

N0rator replied:
Done

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+Use IDE that support development on .NET like [Visual Studio](https://visualstudio.microsoft.com/ru/) or [Intellij Rider](https://www.jetbrains.com/ru-ru/rider/).
+
+### Swagger
+To use Swagger functionality, get on `url:port/Swagger`.
```
Make that exact valid link, e.g.
```suggestion
To use Swagger functionality, get on `https://localhost:3000/Swagger`.
```
Or make port 3000 to be just "port", but default address would be nice as it is expected to be single variant as swagger would be used only for local development I guess.

N0rator replied:
Added valid links

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+### Swagger
+To use Swagger functionality, get on `url:port/Swagger`.
+Swagger is enabled only in **Debug** configuration.
```
And how to enable/disable that?

N0rator replied:
To enable Swagger, just build application in Debug configuration, to disable - in Release configuration.
Application in build directory is built in Release configuration.

Max-Dov commented on WorldProcessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+##  Compiled application
+Latest compiled application is located in `Release build` directory.
+To use it, just run `WorldProcessor.WebApi.exe` executable file and find application address and port in console.
```
Do .dll and other files required for that .exe?
Is it possible to place everything within single .exe, so folder "build" for that branch would have simple single .exe file?

N0rator replied:
It is possible to minimize final application, but some files like `appsettings.json` will stay due to application nature.

I will change current build to minimized

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Dto/NeighborConstraintDto.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace WorldProcessor.Application.Handlers.WorldProcessing.Dto
+{
+    public class NeighborConstraintDto
```
Would be good to have documentation for all these classes, but it's just me. I can forget what code does or what is purpose of a variable after 1 week, so I prefer to document code even I write for myself.

Not push tho, do whatever you feel like. This info in contracts anyway (you can put link in doc for example).

N0rator replied:
Good proposal, but I'd better to leave it as how it done for now and add any documentation in near future while refactoring.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Dto/TileConfigDto.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        public string Id { get; set; }
+
+        public NeighborConstraintDto[]? Neighbors { get; set; }
```
Does `?` means optional (means can be `null`)?
If so, then neighbors is not an optional param, it is required param and always would be an array. Maybe an empty array, but an array nevertheless

N0rator replied:
This was done after testing with using world config example from config-editor's readme file. Unfortunately, aspnet model builder is strict to such cases and will throw validation errors.
I see 2 solutions:
1) Require strict json schema in requests.
2) Initialize this property in class constructor.

I dont like second option, because i think that such logic is not compatible with DTO definition. But i may be wrong and would like to see your opinion.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Dto/TileConfigDto.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public int MutationWeight { get; set; }
+
+        public int MutationChance { get; set; }
```
I would make it `double` for crafting "extremely" rare to mutate tiles with like 0.01% change to mutate.

But we have no actual usecase for it yet, so it's up to you. But you might need to refactor it later when usecase will pop up.

N0rator replied:
Accepted and implemented.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Queries/GenerateWorldQuery.cs:
```diff
@@ -0,0 +1,15 @@
+        public IEnumerable<int>? Dimensions { get; set; }
+
+        public WorldConfigDto WorldConfig { get; init; }
+    }
```
Missing `epoch`.
We may want to create world with not just "0" as starting epoch, but something epic, like 12 or even 2012.

N0rator replied:
Added.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Dto/WorldDto.cs:
```diff
@@ -0,0 +1,17 @@
+
+        public int Seed { get; set; }
+
+        public int Epoch { get; set; }
```
Does `int` allow negative numbers just in case?

N0rator replied:
Yep. I see a your thought. I will add validation for this case.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Mapping/WorldIterationAutoMapperProfile.cs:
```diff
@@ -0,0 +1,169 @@
+
+                if(tilePosition.GetType() != result.Dimensions.GetType())
+                {
+                    throw new Exception(
```
You throw quite a few exceptions, but I don't see them catched anywhere. Would web api be online still after a couple of exceptions?

N0rator replied:
Yes, it will continue to work. Because this application is stateless, the exception will not cause problems. But this is a good point. I will add exception handlers. While they will be simple and will be improved in the near future.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Mapping/WorldIterationAutoMapperProfile.cs:
```diff
@@ -0,0 +1,169 @@
+                if(tilePosition.GetType() != result.Dimensions.GetType())
+                {
+                    throw new Exception(
+                        "Tile position type is not similar to world dimensions type - " +
```
```suggestion
                        "Tile position type is not equal to world dimensions type - " +
```

N0rator replied:
Fixed.

Max-Dov commented on WorldProcessor/WorldProcessor.Application/Handlers/WorldProcessing/Queries/GenerateWorldQueryHandler.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+            var worldConfig = _mapper.Map<WorldConfigDto, WorldConfig>(request.WorldConfig);
+
+            var result = new World();
```
imagine using `var` in 2022 when JS already invented `let` and `const` like 10 years ago

N0rator replied:
While front-end developers are living in 2022, back-end developers are hiting servers by leg to force their work.

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Entities/Tile.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public string Representation { get; set; }
+
+        public int MutationChance { get; set; }
```
🚑
That is 100% `double`. Base mutation chance is `int`, but multipliers are `double` and result would not be `int` for sure.

N0rator replied:
Fixed

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Entities/Tile.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class Tile : ICloneable
+    {
+        public string ConfigId { get; set; }
```
I see all of your configIds in all models have `set`, but they can't be realistically set to anything else once initialized.
Is it a mistake or all is ok?

N0rator replied:
It is not a good practice, yes. But it happened because objects mutates in iteration and generation algorithms. I suggest to fix it later while WorldProcessor enhancement.

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Enums/HexagonalDirection.cs:
```diff
@@ -0,0 +1,12 @@
+        YPlus = 3,
+        ZMinus = 4,
+        XPlus = 5,
+        YMinus = 6,
```
👀 wth

N0rator replied:
unknown slang

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Enums/TetragonalDirection.cs:
```diff
@@ -0,0 +1,14 @@
+        South = 5,
+        SouthWest = 6,
+        West = 7,
+        NorthWest = 8,
```
Neigbor realistically is tile that has common sides with origin.
Hexagonal tile has 6 neighbors and it has 1 common side with everyone to them.

With quads (tetragons) you realistically have only 4 neighbors, because you don't share sides with diagonal neighbors.

And with tiles, directions represent neares neighbor location. That means quad tiles should have only 4 directions to go.

But that thing is up to a debate how we want it to be implemented.

N0rator replied:
Tetragonal world support is abandoned

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/RandomValueGenerationService.cs:
```diff
@@ -0,0 +1,26 @@
+            double iterationMix = Math.Abs(Math.Sin(iteration));
+            var shaking = seedMix + epochMix + coordinatesMix + iterationMix;
+
+            var stringResult = "0," + new string(shaking.ToString().Reverse().Take(..^2).ToArray());
```
that is slow compared to for loop
can you make it via simple `for` or while loop?
i think that would save quite a lot of computing time

something like that
```cpp
        int tempShaking = shaking;
        int reverted = 0;
        while (tempShaking  != 0) {
            int digit = x % 10;
            tempShaking /= 10;
            reverted = reverted * 10 + digit;
        }
```

you can probably measure time improvement or time downgrade yourself.

N0rator replied:
improved

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldGenerationService.cs:
```diff
@@ -0,0 +1,170 @@
+            {
+                ConfigId = tileConfig.Id,
+                BirthEpoch = 0,
+                Representation = "",
```
🔫 👀

N0rator replied:
fixed

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+            return result;
+        }
+
+        private bool IsTileNeedToMutate(Tile tile, IPosition position, TileConfig config, World world)
```
```suggestion
        private bool doesTileNeedToMutate(Tile tile, IPosition position, TileConfig config, World world)
```
(optional)

N0rator replied:
It will stay as it done for now.

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+
+        private bool IsTileNeedToMutate(Tile tile, IPosition position, TileConfig config, World world)
+        {
+            if (config.MinAge == 0 ||
```
Tile needs to mutate if it's too old (tile.age > config.maxAge)
```suggestion
            if (config.MaxAge == 0 ||
```

N0rator replied:
fixed

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+        private bool IsTileNeedToMutate(Tile tile, IPosition position, TileConfig config, World world)
+        {
+            if (config.MinAge == 0 ||
+                tile.BirthEpoch + config.MinAge >= world.Epoch)
```
Я вдруг вспомнил, что можно писать на русском

Это условие не логично.
"Дата рождения сложеная с минимальным порогом жизни" -- втф?
Предлагаю написать "Возраст мира минус дата рождения" -- получается возраст тайла -- "должен быть меньше максимального возраста тайла и больше минимального возраста тайла"
Предлагаю похожий на такое код:
```c#
int tileAge = world.Epoch - tile.BirthEpoch;
if (tileAge > config.MaxAge) return true;
if (tileAge < config.MinAge) return false;
// ... бросок MutationChance
```

N0rator replied:
improved

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+            var mutationCheckRandomValue = _randomValueGenerationService
+                .Generate(
+                    world.Seed,
+                    world.Epoch + 1,
```
```suggestion
                    world.Epoch,
```
Давай не заглядывать в будущее на эту итерацию )

N0rator replied:
fixed

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+                    position, 0
+                    );
+
+            return tile.MutationChance >= mutationCheckRandomValue;
```
MutationChance же число от 0 до 100 фактически
mutationCheckRandomValue это от 0 до 1

ошибаюсь?

N0rator replied:
fixed

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/RandomValueGenerationService.cs:
```diff
@@ -0,0 +1,26 @@
+        public RandomValueGenerationService()
+        { }
+
+        public double Generate(int seed, int epoch, IPosition position, int iteration = 0)
```
Рекомендую сделать это сервисом (штука с состоянием), что будет запоминать количество вызовов итераций для seed+int+position тульпы.

Это исключит человеческую ошибку, когда будет прокинут неправильный номер итерации на какой-нибудь ивент, что может привести либо к неточным числам, либо вообще к одинаковым событиям на разные действия.

N0rator replied:
Пожалуй пусть это останется так как есть. Разве что я добавлю enum для аргумента iteration.

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldGenerationService.cs:
```diff
@@ -0,0 +1,170 @@
+            };
+        }
+
+        private IEnumerable<TileConfig> GetAvailableTilesForPosition(
```
@Max-Dov почитать завтра на свежую голову

Max-Dov commented on WorldProcessor/WorldProcessor.Core/Services/WorldIterationService.cs:
```diff
@@ -0,0 +1,265 @@
+                TileShape = world.TileShape,
+                Epoch = world.Epoch + 1,
+                Seed = world.Seed,
+                Tiles = world.Tiles.ToDictionary(
```
Логически новая итерация мира не предполагает мутацию предыдущего, а создание новой версии, с нуля.

Это я к тому, что тайлы я бы не копировал, а добавлял по мере понимания, какой тайл должен быть на этой координате.

Иначе получается неочевидный код. А неочевидный код часто приводит к багам.

Max-Dov commented on WorldProcessor/WorldProcessor.Core/ValueTypes/Position3.cs:
```diff
@@ -0,0 +1,160 @@
+
+namespace WorldProcessor.Core.ValueTypes
+{
+    public struct Position3 : IPosition
```
Странный нейминг Position. Я бы назвал это минимум координатой, максимум вектором. (Не меняй если не хочешь)

Max-Dov commented on WorldProcessor/WorldProcessor.Core/ValueTypes/Position2.cs:
```diff
@@ -0,0 +1,139 @@
+
+namespace WorldProcessor.Core.ValueTypes
+{
+    public struct Position2 : IPosition
```
вот ты запарился еще и с quadами возиться

Max-Dov commented on notes/contracts.md:
```diff
@@ -207,7 +207,7 @@ Misc
  * @returns next or first world iteration
  */
-type iterateWorld = (worldConfig: WorldConfig, world?: World) => World
+type iterateWorld = (worldConfig: WorldConfig, world: World) => World
```
Добавь определение generateWorld, пожалуйста. 👉 👈

Max-Dov reviewed (changes requested):
Это жестко, чутка поправить там и сям надо 🤷

N0rator committed:
world-processor: big initial commit

That is done so, because we tried to remove large executable from git repo history. Hope we successed.
This commit is a squash of this list:
* rename directories
* update README.md file
* update README.md file according to comments in pull request.
* another update for README.md file
* add exception handlers
* fix lexical mistake
* change MutationChance type, add functionality to set initial epoch for world generation
* fix generate world request body description
* add representation mutation
* improve random value generation speed
* refactor check for tile mutation need
* fixed roll in check for tile mutatuon need
* remove asynchronious in world iteration service
* code improvements
* add input validation
* fix iteration, improve generation
* fix validation

N0rator committed:
world-processor: reach functionality level of node.js version

Max-Dov commented:
Переоткрой когда доведешь до 1.0
